src/main/java/com/zendesk/maxwell/filtering/FilterColumnPattern.java

```diff
@@ -0,0 +1,71 @@
+package com.zendesk.maxwell.filtering;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+public class FilterColumnPattern extends FilterPattern {
+	private static final Logger LOGGER = LoggerFactory.getLogger(FilterColumnPattern.class);
+	private final String columnName;
+	private final Pattern columnPattern;
+	private final boolean columnPatternIsNull;
+
+	public FilterColumnPattern(
+		FilterPatternType type,
+		Pattern dbPattern,
+		Pattern tablePattern,
+		String columnName,
+		Pattern columnPattern
+	) {
+		super(type, dbPattern, tablePattern);
+		this.columnName = columnName;
+		this.columnPattern = columnPattern;
+		this.columnPatternIsNull = "^null$".equals(columnPattern.toString().toLowerCase());
+	}
+
+	@Override
+	public void match(String database, String table, FilterResult match) { }
```

Comment:
This line and [this](https://github.com/zendesk/maxwell/pull/1010/files#diff-0a3ff09dd0063fec1059f7538d686c75R28) make me think whether `FilterColumnPattern` _is_a_ `FilterPattern`. Strictly speak, whether it _is_a_ `FilterDbTablePattern`. But I don't have a better way to address it. Maybe it's because what you said that we do a two-pass filter and we are checking db and table twice (in both filter patterns).

Reply:
I'm thinking out loud so bear with me if it doesn't make sense.

Is it possible to consolidate table and column patterns so eventually the table pattern _has_a_ column pattern (can only be done after all the rules have been parsed). Because we already have `couldIncludeFromColumnFilters` method so I assume the table pattern could be aware if it has an associated column pattern. 

It might break the encapsulation of `shouldOutputEvent` b/c it will return the table pattern(s) that are evaluated as true (by extending `FilterResult.patterns`), instead of only the boolean (`FilterResult.include`) at the moment. Then `shouldOutputRowMap` will use the associated column patterns (if there is) to filter. It will make the filtering kinda stateful compared to the current completely stateless impl but it won't check db/table the second time on each `RowMap`. 

Maybe it's somehow similar to your 1st alternative? Would it work at all?

Reply:
I don't think a table pattern can strictly has-a column pattern, since rules could look like:

```
   exclude: db.*
   include: db.included_table
   exclude: db.included_table.col = reject_value_1
   exclude: db.included_table.col = reject_value_2
```

unless I'm not quite getting your idea.

Reply:
Right, I wasn't clear enough. I meant a table column _has_a_ optional column pattern(s). Anyways, I tried it out and don't think it's practical in terms of associating the optional column pattern with the table patten right after the parsing. If we can, it will make it much easier later on.

I put together some hacky code 
[7655c5f](https://github.com/zendesk/maxwell/commit/7655c5fe922dba3bfa4fbe4b9eb2e9cfeb8e695f) to demonstrate the second part of my idea. To summarise:

- separate the two patterns completely (they might both extend `AbstractPattern` to reuse some code).
- Use `FilterResult` to hold the state, i.e. tablePatterns and columnPatterns that match the db and table in `shouldOutputEvent` so in `shouldOutputRowMap` it doesn't have to check db/table any more.

The downside is it breaks the encapsulation of the filter. The upside is to remove the duplicate db/table check.  

BTW, I wonder why we need those static methods in `Filter`, e.g. `includes`s